packages/realm-react/src/useAuth.tsx

```diff
@@ -0,0 +1,196 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { useCallback } from "react";
+import { useApp, useAuthResult } from "./AppProvider";
+import { AuthError, AuthOperationName, AuthResult, OperationState } from "./types";
+import { Realm } from "realm";
+import { useAuthOperation } from "./useAuthOperation";
+
+/**
+ * Hook providing operations and corresponding state for authenticating with a
+ * Realm app.
+ *
+ * The {@link AuthResult} states returned from this hook are "global" for all
+ * components under a given AppProvider, as only one operation can be in progress
+ * at a given time (i.e. we will store the states on the context). This means that,
+ * for example, multiple components can use the `useAuth` hook to access
+ * `loginResult.pending` to render a spinner when login is in progress, without
+ * needing to pass that state around or store it somewhere global in their app
+ * code.
+ */
+interface UseAuth {
+  /**
+   * Log in with a {@link Realm.Credentials} instance. This allows login with any
+   * authentication mechanism supported by Realm.
+   * If this is called when a user is currently logged in, it will switch the user.
+   *
+   * @returns A `Realm.User` instance for the logged in user.
+   */
+  logIn(credentials: Realm.Credentials): Promise<Realm.User | void>;
+
+  /**
+   * Log in with the Anonymous authentication provider.
+   *
+   * @returns A `Realm.User` instance for the logged in user.
+   * @see https://www.mongodb.com/docs/atlas/app-services/authentication/anonymous/
+   */
+  logInWithAnonymous(): Promise<Realm.User | void>;
+
+  /**
+   * Log in with an API key.
+   *
+   * @returns A `Realm.User` instance for the logged in user.
+   * @see https://www.mongodb.com/docs/atlas/app-services/authentication/api-key/
+   */
+  logInWithApiKey(key: string): Promise<Realm.User | void>;
+
+  /**
+   * Log in with Email / Password.
+   *
+   * @returns A `Realm.User` instance for the logged in user.
+   * @see https://www.mongodb.com/docs/atlas/app-services/authentication/email-password/
+   */
+  logInWithEmailPassword(params: { email: string; password: string }): Promise<Realm.User | void>;
+
+  /**
+   * Log in with a JSON Web Token (JWT).
+   *
+   * @returns A `Realm.User` instance for the logged in user.
+   * @see https://www.mongodb.com/docs/atlas/app-services/authentication/custom-jwt/
+   */
+  logInWithJWT(token: string): Promise<Realm.User | void>;
+
+  /**
+   * Log in with Google.
+   *
+   * @returns A `Realm.User` instance for the logged in user.
+   * @see https://www.mongodb.com/docs/atlas/app-services/authentication/google/
+   */
+  logInWithGoogle(params: { idToken: string } | { authCode: string }): Promise<Realm.User | void>;
```

Comment:
I am mirroring what is set at the SDK level.  Perhaps I should think of a way to do this generically, so that if at some point the underlying Credential changes, the hooks would still function.  I do agree that the naming is inconsistent and could be misleading.

Reply:
Ah okay gotcha!